rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,756 @@
+  call void @llvm.lifetime.end.p0i8(i64 40, i8* nonnull %52), !noalias !99611
+  br label %47
+}
+```
+
+Note the batched stack allocations, inlining of function calls and consolidation
+of control flow.
+
+## Rationale
+
+As long as the single-core performance of VRL is the bottleneck of a topology,
+general performance improvements to VRL are extremely valuable as they equate to
+an equally sized performance improvement to the entire topology.
+
+We want to live up to the
+[performance guarantees](https://github.com/vectordotdev/vector/blob/f1404bea186ba83c4426a32bbef3f633c17cf4d2/website/cue/reference/remap/features/compilation.cue#L4-L8)
+outlined in VRL's list of features.
+
+Every VRL program benefits from the reduced runtime overhead, without us needing
+to optimize any specific use case. Consistent execution speed is important to
+build trust in the language.
+
+Being able to execute our log transformation DSL at speeds which would otherwise
+only be attainable by hand-writing Rust programs will strengthen a key value
+proposition of Vector: best-in-class performance.
+
+## Drawbacks
+
+By generating machine code via LLVM, we are no longer (largely) immune to memory
+violations. To reduce the error surface as much as possible, we employ industry
+practices like fuzz-testing VRL programs, static analysis via LLVM and rely on
+the Rust compiler for any non-trivial code fragments. That being said, memory
+safety is an illusion - with the difference that we are now exposing ourselves
+to being partly responsible for maintaining invariants instead of being able to
+defer to a third party for correctness. We still provide an inherently memory
+safe language to the user.
+
+Producing LLVM bitcode for Rust's `std` library is guarded behind the
+[`-Z build-std`](https://doc.rust-lang.org/cargo/reference/unstable.html#build-std)
+flag and only available on the nightly compiler toolchain. We need `std` to
+fully link our precompiled LLVM bitcode. It's possible to circumvent the nightly
+requirement by setting the `RUSTC_BOOTSTRAP=1` environment variable, such that
+we have a `std` that is built using the same Rust and LLVM version as Vector. We
+isolate the usage of this hack by building a separate crate with `std` only, and
+link it to the library bitcode in a build step.
+
+Statically linking LLVM to the Vector binary adds roughly 9MB, additionally to
+precompiled bitcode that needs to be included with the binary. If this is a
+concern, we can consider shipping binaries with the LLVM feature disabled.
+
+While LLVM is a highly used framework within the industry, working with it
+requires rather specialized knowledge about compiler construction. However,
+there exists plenty of publicly accessible material for code generation using
+LLVM, some of which I linked to [above](#introduction-to-llvm). In addition,
+there should be great focus to document this part of the code base
+extraordinarily well.
```

Comment:
```suggestion
there should be great focus to document and test this part of the code base
extraordinarily well.
```

Reply:
Adressed in 2df80e5c35211c6fae24313d31f84a99f7dbd6e1.